include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,301 @@
+					{
+						*(currentColumnCache + channel) += *(mainRowCacheScratch + channel) += decodeBuffer[channel];
+						*(finalPixel + channel) = *(currentColumnCache + channel) - (ExclusiveMode ? decodeBuffer[channel] : 0);
+					}
+
+					memcpy(scratchMemory + outDataAdressOffsetScratch + layersOffset, finalPixel, state->mainRowScratchCacheByteSize);
+
+					if (newReadBlockPos.x == trueOutExtent.width - 1) // reset row cache when changing y
+					{
+						if (newReadBlockPos.y == 0) // sum values in (maxX, 0, z) and remove values from column cache
+						{ 
+							memcpy(totalImageCacheScratch + (newReadBlockPos.w * state->extent.depth + newReadBlockPos.z) * state->mainRowScratchCacheByteSize, finalPixel, state->mainRowScratchCacheByteSize);
+							memset(mainColumnCacheScratch, 0, state->mainColumnScratchCacheByteSize);
+						}
+
+						memset(mainRowCacheScratch, 0, state->mainRowScratchCacheByteSize);
+					}
+				};
+
+				{
+					core::vector3du32_SIMD localCoord;
+					for (auto& w = localCoord[3] = 0u; w < arrayLayers; ++w)
+						for (auto& z = localCoord[2] = 0u; z < trueOutExtent.depth; ++z)
+							for (auto& y = localCoord[1] = 0u; y < trueOutExtent.height; ++y)
+								for (auto& x = localCoord[0] = 0u; x < trueOutExtent.width; ++x)
+									sum(core::vectorSIMDu32( x, y, z, w ));
+				}
+
+				auto normalizeScratch = [&]()
+				{
+					core::vector3du32_SIMD localCoord;
+					for (auto& w = localCoord[3] = 0u; w < arrayLayers; ++w)
+						for (auto& z = localCoord[2] = 0u; z < trueOutExtent.depth; ++z)
+							for (auto& y = localCoord[1] = 0u; y < trueOutExtent.height; ++y)
+								for (auto& x = localCoord[0] = 0u; x < trueOutExtent.width; ++x)
+								{
+									const auto scratchInPlaneOffset = ((localCoord.z * state->extent.height + localCoord.y) * state->extent.width + localCoord.x);
+									const auto outDataAdressOffsetScratch = scratchInPlaneOffset * currentChannelCount;
+
+									auto* entryScratchAdress = scratchMemory + outDataAdressOffsetScratch + singleScratchLayerOffset * w;
+									auto* entryTotalImageCacheValues = reinterpret_cast<decodeType*>(totalImageCacheScratch + (localCoord.w * state->extent.depth + localCoord.z) * state->mainRowScratchCacheByteSize);
+
+									for (uint8_t channel = 0; channel < currentChannelCount; ++channel)
+										entryScratchAdress[channel] /= entryTotalImageCacheValues[channel];
+								}
+				};
+
+				if (state->normalizeImageByTotalSATValues)
+					if (!asset::isNormalizedFormat(inFormat))
+						normalizeScratch();
+
+				auto perOutputRegionEncode = [&](const CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
+				{
+					auto encode = [&](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+					{
+						const auto scratchInPlaneOffset = ((readBlockPos.z * state->extent.height + readBlockPos.y) * state->extent.width + readBlockPos.x);
```

Comment:
you do realize that `readBlockPos` is a global coordinate in the texture, not local w.r.t. your output offset and hence your scratch memory?

Reply:
since scratch covers my full extent I can use it for sure because your common execute lambda will perform clipping to an particular extent with some offset if given, and knowing the scratch extent in advance I can specify `scratchInPlaneOffset` to later user it to compute real byte offsets (knowing channel amount, the rest is on pointer arithmetic side(

Reply:
and it's locaal https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/irr/asset/filters/CBasicImageFilterCommon.h#L49

Reply:
Nope its the second parameter of this lambda, and it gets fed with `localCoord+trueOffset`, so its global (relative to mip levels origin) in its mip level

Reply:
I think we're clear now?

Reply:
I think yeah, it will be documented soon by the way:)